### PR TITLE
Feature/csl691 refactor relay

### PR DESCRIPTION
**Pull request by georgeee**

### CODE REVIEW

File: src/Pos/Communication/Relay.hs [hunk 1]
```diff
@@ -122,104 +130,214 @@ handleReqL
...
+  \sendActions -> handleAll handleWE $ do
+    queue <- ncInvPropagationQueue <$> getNodeContext
+    forever $ do
+        inv <- atomically $ readTBQueue queue
+        case inv of
+            SomeInvMsg (i@(Left (InvMsg{..}))) -> do
+                logDebug $
+                    sformat ("Propagation data with keys: "%listJson%" and tag: "%build) imKeys imTag
+                converseToNeighbors sendActions (convHandler i)
+            SomeInvMsg (Right _) -> throw InvalidPropagationElement
```
Comment (georgeee) [inline src/Pos/Communication/Relay.hs]: No error throwing!! Just log. As it's written now, worker would finish, which shouldn't happen.

File: src/Pos/Communication/Relay.hs [hunk 2]
```diff
@@ -122,104 +130,214 @@ handleReqL
...
      , Bi (ReqMsg key tag)
      , Bi (DataMsg contents)
+     , Bi (InvOrData tag key contents)
+     , MessagePart contents
+     , MessagePart tag
      , Relay m tag key contents
+     , Mockable Throw m
      , WithLogger m
      , WorkMode ssc m
        -- [CSL-659] remove after rewriting to conv
```
Comment (georgeee) [inline src/Pos/Communication/Relay.hs]: And here

File: src/Pos/Communication/Relay.hs [hunk 3]
```diff
@@ -122,104 +130,214 @@ handleReqL
...
-reqCatchType :: RelayProxy key tag contents -> ReqMsg key tag -> ()
-reqCatchType _ _ = ()
-reqMsgProxy :: RelayProxy key tag contents -> Proxy (ReqMsg key tag)
-reqMsgProxy _ = Proxy
+    handleDataLDo dmKey = do
+        shouldPropagate <- ncPropagation <$> getNodeContext
+        if shouldPropagate then do
+            tag <- contentsToTag dmContents
+            let inv :: InvOrData tag key contents
+                inv = Left $ InvMsg tag (one dmKey)
```
Comment (georgeee) [inline src/Pos/Communication/Relay.hs]: I don't get it? Why put `Left` to queue and explicitly fail on receiving `Right` in worker?

File: src/Pos/Communication/Relay.hs [hunk 4]
```diff
@@ -93,27 +99,29 @@ handleInvL
...
 
 handleReqL
-    :: ( Bi (ReqMsg key tag)
-       , Bi (DataMsg contents)
+    :: forall key tag contents m .
+       ( Bi (ReqMsg key tag)
+       , Bi (InvOrData tag key contents)
        , Relay m tag key contents
        , MinWorkMode m
        -- [CSL-659] remove after rewriting to conv
```
Comment (georgeee) [inline src/Pos/Communication/Relay.hs]: ```Bi NOP, Message NOP``` to be removed

File: src/Pos/Communication/Message.hs [hunk 5]
```diff
@@ -55,13 +57,20 @@ instance Message (MsgBlock s ssc) where
...
         tagM _ = Proxy
-    formatMessage _ = "Inventory"
+
+        contentsM :: Proxy (InvOrData tag keys contents)
+                  -> Proxy contents
+        contentsM _ = Proxy
+    formatMessage _ = "Inventory/Data"
 
 instance (MessagePart tag) =>
          Message (ReqMsg key tag) where
```
Comment (pva701) [inline src/Pos/Communication/Message.hs]: no, we need `Message (InvOrData tag key contents)`

File: src/Pos/Ssc/GodTossing/Workers.hs [hunk 6]
```diff
@@ -122,10 +124,12 @@ checkNSendOurCert sendActions = do
             ourVssCertificate <- getOurVssCertificate
             let contents = MCVssCertificate ourVssCertificate
             sscProcessOurMessage contents
-            let msg = DataMsg contents
-            sendToNeighbors sendActions msg
-            logDebug "Announced our VssCertificate."
-    (_, ourId) <- getOurPkAndId
+            let msg = InvMsg VssCertificateMsg (one ourId)
+            -- [CSL-245]: do not catch all, catch something more concrete.
+            (converseToNeighbors sendActions (sendInv msg) >>
```
Comment (georgeee) [inline src/Pos/Ssc/GodTossing/Workers.hs]: Why only `Inv` message, not `invReqDataFlow`?

File: src/Pos/Ssc/GodTossing/Workers.hs [hunk 7]
```diff
@@ -264,9 +268,18 @@ sendOurData sendActions msgTag epoch slMultiplier ourId = do
     waitUntilSend msgTag epoch slMultiplier
     logInfo $ sformat ("Announcing our "%build) msgTag
     let msg = InvMsg {imTag = msgTag, imKeys = one ourId}
-    sendToNeighbors sendActions msg
+    converseToNeighbors sendActions (sendInv msg)
```
Comment (georgeee) [inline src/Pos/Ssc/GodTossing/Workers.hs]: Why only Inv message, not invReqDataFlow?

**Pull request merged**